pkg/controller/daemon/daemoncontroller.go

```diff
@@ -501,7 +514,11 @@ func (dsc *DaemonSetsController) resolveControllerRef(namespace string, controll
 		// ControllerRef points to.
 		return nil
 	}
-	return ds
+	cp, err := api.Scheme.DeepCopy(ds)
+	if err != nil {
+		return nil
+	}
+	return cp.(*extensions.DaemonSet)
```

Comment:
The result of this function gets passed to the function that updates the DeletionTimestamp. To make sure all DaemonSet pointers are safe, I made deep copies at the points where they are retrieved from the cache.